### PR TITLE
Delete ReadyPackage

### PR DESCRIPTION
**Pull request by ezyang**
Delete ReadyPackage.

Previously, ReadyPackage was a ConfiguredPackage elaborated with
a dependencies data structure which had InstalledPackageInfo
rather than ConfiguredId.  Well, it turned out that we only
used the data from ConfiguredId!  So ReadyPackage is completely
useless.

I retained some type synonyms to avoid changing type signatures,
and as a reminder that not all ConfiguredPackages can be built,
only the ready ones.

There is also some extra documentation in the rest of the patchset.

### CODE REVIEW

**Comment by 23Skidoo** [comment]
Seems to fail on Travis/AppVeyor.

**Comment by 23Skidoo** [comment]
> Well, it turned out that we only used the data from ConfiguredId! So ReadyPackage is completely useless.

Can't we fix this by making `ReadyPackage` a `newtype` then? I still like the `ReadyPackage`/`ConfiguredPackage` distinction, it makes the intent clearer.

**Comment by ezyang** [comment]
I don't think the newtype would add much (besides line noise), but if you insist, I can reintroduce a newtype.

File: cabal-install/Distribution/Client/Types.hs [hunk 1]
```diff
@@ -163,29 +163,7 @@ instance PackageFixedDeps (ConfiguredPackage loc) where
...
-         PackageFixedDeps (GenericReadyPackage srcpkg ipkg) where
-  depends (ReadyPackage _ deps) = fmap (map installedUnitId) deps
-
-instance HasUnitId srcpkg =>
-         HasUnitId (GenericReadyPackage srcpkg ipkg) where
-  installedUnitId (ReadyPackage pkg _) = installedUnitId pkg
-
-instance (Binary srcpkg, Binary ipkg) => Binary (GenericReadyPackage srcpkg ipkg)
-
+type ReadyPackage = ConfiguredPackage UnresolvedPkgLoc
```
Comment (23Skidoo) [inline cabal-install/Distribution/Client/Types.hs]: Please keep the Haddock comment.

**Comment by 23Skidoo** [comment]
Please do. Since we're not removing the `ReadyPackage` type completely, I think it's better to get some additional type safety.

**Comment by ezyang** [comment]
OK, newtyped.

**Comment by 23Skidoo** [comment]
Thanks! Looks good to go now.

**Comment by ezyang** [comment]
I'll wait for it to green.

**Comment by ezyang** [comment]
I'm really confused: the error message in the Travis build appears to be for a stale version of the commit.

**Pull request closed without merging**